components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -396,12 +399,47 @@ public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices,
             _agentConfig.TriggerConfigUpdate(this, new EventArgs());
         }
 
+        private static void AdjustResponse(ref List<string> statusResponse) {
+            byte[] result;
+            int maxIndex = statusResponse.Count();
+            string resultString = string.Empty;
+            while (true) {
+                resultString = JsonConvert.SerializeObject(statusResponse.GetRange(0, maxIndex));
+                result = Encoding.UTF8.GetBytes(resultString);
+                if (result.Length > MaxResponsePayloadLength) {
+                    maxIndex /= 2;
+                    continue;
+                }
+                else {
+                    break;
+                }
+            }
+            if (maxIndex != statusResponse.Count()) {
+                statusResponse.RemoveRange(maxIndex, statusResponse.Count() - maxIndex);
+                statusResponse.Add("Results have been cropped due to package size limitations.");
+            }
+        }
+        private MethodResponse BuildResponse(List<string> statusResponse, HttpStatusCode statusCode, string prefix) {
+            AdjustResponse(ref statusResponse);
+            string resultString = JsonConvert.SerializeObject(statusResponse);
+            byte[] result = Encoding.UTF8.GetBytes(resultString);
+            if (result.Length > MaxResponsePayloadLength) {
+                _logger.Error($"{prefix} Response size is too long");
+                Array.Resize(ref result, result.Length > MaxResponsePayloadLength ? MaxResponsePayloadLength : result.Length);
+            }
+            MethodResponse methodResponse = new MethodResponse(result, (int)statusCode);
+            _logger.Information($"{prefix} completed with result {statusCode}");
+            return methodResponse;
+        }
+
         /// <inheritdoc/>
-        public async Task<List<string>> PublishNodesAsync(PublishedNodesEntryModel request, CancellationToken ct = default) {
+        public async Task<MethodResponse> PublishNodesAsync(PublishedNodesEntryModel request, CancellationToken ct = default) {
```

Comment:
@monikavar, please do not use `Microsoft.Azure.Devices.Client` specific `MethodResponse` here. That conversion should be done automatically by the controller wrapper. We do not want to have dependency on `Microsoft.Azure.Devices.Client` here. If you need to add new field or modify existing ones, please do that in `PublishedNodesResponseApiModel`. Erroneous cases should be propagated up via exceptions. There is a mechanism of translating them into corresponding HTTP status codes up the stack.

Reply:
+1
let's introduce for each direct method a XxxResponseModel class here as well as XxxResponseApiModel to properly control and encode the payload